From bb925ef61dae8ac219d897b8450b7d681208804f Mon Sep 17 00:00:00 2001
From: timor <timor.dd@googlemail.com>
Date: Sun, 13 Dec 2020 18:46:19 +0100
Subject: [PATCH 08/11] Better support for declarative spacemacs

Sets a bunch of variables based on env vars in `core-load-paths.el`, which are
used in several places to change loading of dotfiles, custom-files and elpa
build directories. (see Readme in overlay for details)
---
 core/core-configuration-layer.el | 14 +++++++------
 core/core-custom-settings.el     | 11 ++++++++++-
 core/core-dotspacemacs.el        | 13 ++++++++----
 core/core-load-paths.el          |  5 +++++
 core/core-spacemacs.el           | 34 ++++++++++++++++++--------------
 5 files changed, 51 insertions(+), 26 deletions(-)

diff --git a/core/core-configuration-layer.el b/core/core-configuration-layer.el
index f4014accc..608d32c05 100644
--- a/core/core-configuration-layer.el
+++ b/core/core-configuration-layer.el
@@ -113,12 +113,14 @@ subdirectory of ROOT is used."
     configuration-layer-elpa-subdirectory
     (if (not dotspacemacs-elpa-subdirectory)
         root
-      (let ((subdir (if (eq 'emacs-version dotspacemacs-elpa-subdirectory)
-                        (format "%d%s%d"
-                                emacs-major-version
-                                version-separator
-                                emacs-minor-version)
-                      (eval dotspacemacs-elpa-subdirectory))))
+      (let ((subdir (if spacemacs-nix-declared-p
+                        spacemacs-nix-suffix
+                      (if (eq 'emacs-version dotspacemacs-elpa-subdirectory)
+                          (format "%d%s%d"
+                                  emacs-major-version
+                                  version-separator
+                                  emacs-minor-version)
+                        (eval dotspacemacs-elpa-subdirectory)))))
         (expand-file-name subdir root))))))
 
 (defun configuration-layer/get-elpa-package-install-directory (pkg)
diff --git a/core/core-custom-settings.el b/core/core-custom-settings.el
index 98ae1b0ab..a0bd9a64f 100644
--- a/core/core-custom-settings.el
+++ b/core/core-custom-settings.el
@@ -23,7 +23,7 @@
 (require 'core-dotspacemacs)
 
 (defvar spacemacs--custom-file (concat spacemacs-cache-directory
-                                       ".custom-settings"))
+                                       ".custom-settings" spacemacs-nix-suffix))
 
 (defun spacemacs/initialize-custom-file ()
   "Initialize the custom file.
@@ -44,6 +44,15 @@ complete that part see `spacemacs/initialize-custom-file-sync'."
 ;; function `dotspacemacs/emacs-custom-settings'.
 ;; Do not alter this file, use Emacs customize interface instead.\n\n")))))
 
+(defun spacemacs/initialize-nix-custom-file ()
+  "If declarative nix Spacemacs is used, set up custom value of
+  `custom-file', bypassing regular behavior."
+  (if spacemacs-nix-declared-p
+      (progn
+        (setq custom-file (concat spacemacs--custom-file "-" spacemacs-nix-suffix))
+        (when (file-exists-p custom-file) (load custom-file)))
+    (spacemacs/initialize-custom-file)))
+
 (defun spacemacs/initialize-custom-file-sync ()
   "Initialize syncing of the custom file to the dotfile."
   (when (string-equal custom-file spacemacs--custom-file)
diff --git a/core/core-dotspacemacs.el b/core/core-dotspacemacs.el
index 307d3cc4b..09433c598 100644
--- a/core/core-dotspacemacs.el
+++ b/core/core-dotspacemacs.el
@@ -55,11 +55,16 @@ their configuration.")
       (if (file-regular-p spacemacs-init)
           spacemacs-init
         (let ((fallback-init (expand-file-name ".spacemacs.d/init.el"
-                                               user-home-directory)))
-          (if (file-regular-p fallback-init)
-              fallback-init
-            spacemacs-init))))
+                                               user-home-directory))
+              (default-nix-init (when spacemacs-nix-declared-p
+                                  (expand-file-name
+                                   (concat ".spacemacs-" (getenv "NIX_SPACEMACS_SUFFIX"))
+                                   dotspacemacs-directory))))
+          (cond  (default-nix-init)
+                 ((file-regular-p fallback-init) fallback-init)
+                 (t spacemacs-init)))))
   "Filepath to Spacemacs configuration file (defaults to ~/.spacemacs).
+- If declarative spacemacs is built with Nix given a dotfile, get proper generated one from environment
 - If environment variable SPACEMACSDIR is set and SPACEMACSDIR/init.el
   exists, use that value.
 - Otherwise use ~/.spacemacs if it exists.
diff --git a/core/core-load-paths.el b/core/core-load-paths.el
index abaa2a226..cd25d28f8 100644
--- a/core/core-load-paths.el
+++ b/core/core-load-paths.el
@@ -37,6 +37,11 @@
   (concat (file-name-directory (or load-file-name buffer-file-name)) "../")
   "Spacemacs start directory.")
 
+;; Support for Nix declarative Spacemacs
+(defconst spacemacs-nix-declared-p (getenv "NIX_SPACEMACS_SUFFIX"))
+(defconst spacemacs-nix-suffix (or (getenv "NIX_SPACEMACS_SUFFIX") ""))
+(defconst spacemacs-nix-dotfile-path (getenv "NIX_DOTSPACEMACS"))
+
 ;; ~/.emacs.d/assets
 (defconst spacemacs-assets-directory
   (concat spacemacs-start-directory "assets/")
diff --git a/core/core-spacemacs.el b/core/core-spacemacs.el
index bf5501d6e..9f33991e9 100644
--- a/core/core-spacemacs.el
+++ b/core/core-spacemacs.el
@@ -70,6 +70,22 @@
   "Whether or not spacemacs has finished initializing by completing
 the final step of executing code in `emacs-startup-hook'.")
 
+(defun spacemacs//init-nix-dotfile ()
+  "Custom dotfile handling if provided by declarative nix build."
+
+  (when spacemacs-nix-declared-p
+    ;; dotspacemacs-filepath is store-path dependent location in this case
+    (if (file-exists-p dotspacemacs-filepath)
+        (let ((buf1 (find-file-noselect dotspacemacs-filepath nil t))
+              (buf2 (find-file-noselect spacemacs-nix-dotfile-path nil t)))
+          (when (not (= 0 (compare-buffer-substrings buf1 nil nil buf2 nil nil)))
+            (warn "Eval-time dotfile %s differs from loaded dotfile %s.
+          Rebuild spacemacs with nix to resolve." spacemacs-nix-dotfile-path dotspacemacs-filepath))
+          (kill-buffer buf1)
+          (kill-buffer buf2))
+      (copy-file spacemacs-nix-dotfile-path dotspacemacs-filepath)
+      (set-file-modes dotspacemacs-filepath #o644))))
+
 (defun spacemacs/init ()
   "Perform startup initialization."
   (setq command-line-args (spacemacs//parse-command-line command-line-args))
@@ -95,20 +111,8 @@ the final step of executing code in `emacs-startup-hook'.")
   ;; Nixos adjustment, allows loading a different dotfile then the default one,
   ;; but only during initialization. Subsequent uses of the dotfile location
   ;; should point to the one the user is supposed to edit.
-  (let* ((dotspacemacs-filepath-orig dotspacemacs-filepath)
-         (dotspacemacs-filepath
-          (cond ((getenv "NIX_DOTSPACEMACS"))
-                (t dotspacemacs-filepath))))
-    (when (not (string= dotspacemacs-filepath-orig
-                      dotspacemacs-filepath))
-      (let ((buf1 (find-file-noselect dotspacemacs-filepath nil t))
-            (buf2 (find-file-noselect dotspacemacs-filepath-orig nil t)))
-        (when (not (= 0 (compare-buffer-substrings buf1 nil nil buf2 nil nil)))
-          (warn "Custom dotfile %s differs from loaded dotfile %s.
-          Rebuild spacemacs with nix to resolve." dotspacemacs-filepath-orig dotspacemacs-filepath))
-        (kill-buffer buf1)
-        (kill-buffer buf2)))
-    (dotspacemacs/load-file))
+  (spacemacs//init-nix-dotfile)
+  (dotspacemacs/load-file)
   (dotspacemacs|call-func dotspacemacs/init "Calling dotfile init...")
   (when dotspacemacs-undecorated-at-startup
     ;; this should be called before toggle-frame-maximized
@@ -134,7 +138,7 @@ the final step of executing code in `emacs-startup-hook'.")
   ;; effective ones.
   ;; Note: Loading custom-settings twice is not ideal since they can have side
   ;; effects! Maybe an inhibit variable in Emacs can suppress these side effects?
-  (spacemacs/initialize-custom-file)
+  (spacemacs/initialize-nix-custom-file)
   ;; Commenting the first load although it is mentioned above that we must do it
   ;; I don't recall why we must load the custom settings twice and my experiment
   ;; seems to show that we don't need this double loading process anymore.
-- 
2.42.0

